provisioning/baremetal_config.go

```diff
@@ -75,56 +78,129 @@ func getProvisioningNetworkMode(prov *metal3iov1alpha1.Provisioning) metal3iov1a
+		"ProvisioningDHCPRange":     prov.Spec.ProvisioningDHCPRange,
+		"ProvisioningOSDownloadURL": prov.Spec.ProvisioningOSDownloadURL,
 	}
-	return nil
+
+	return validateProvisioningConfig(provisionConfig)
 }
 
 func validateUnmanagedConfig(prov *metal3iov1alpha1.Provisioning) error {
-	for _, toTest := range []struct {
-		Name  string
-		Value string
-	}{
-
-		{Name: "ProvisioningInterface", Value: prov.Spec.ProvisioningInterface},
-		{Name: "ProvisioningIP", Value: prov.Spec.ProvisioningIP},
-		{Name: "ProvisioningNetworkCIDR", Value: prov.Spec.ProvisioningNetworkCIDR},
-		{Name: "ProvisioningOSDownloadURL", Value: prov.Spec.ProvisioningOSDownloadURL},
-	} {
-		if toTest.Value == "" {
-			return fmt.Errorf("%s is required but is empty", toTest.Name)
+	provisionConfig := map[string]string{
+		"ProvisioningInterface":     prov.Spec.ProvisioningInterface,
+		"ProvisioningIP":            prov.Spec.ProvisioningIP,
+		"ProvisioningNetworkCIDR":   prov.Spec.ProvisioningNetworkCIDR,
+		"ProvisioningOSDownloadURL": prov.Spec.ProvisioningOSDownloadURL,
+	}
+
+	return validateProvisioningConfig(provisionConfig)
+}
+
+func validateDisabledConfig(prov *metal3iov1alpha1.Provisioning) error {
+	provisionConfig := map[string]string{
+		"ProvisioningIP":            prov.Spec.ProvisioningIP,
+		"ProvisioningNetworkCIDR":   prov.Spec.ProvisioningNetworkCIDR,
+		"ProvisioningOSDownloadURL": prov.Spec.ProvisioningOSDownloadURL,
+	}
+
+	return validateProvisioningConfig(provisionConfig)
+}
+
+func validateProvisioningOSDownloadURL(uri string) error {
+	parsedURL, err := url.ParseRequestURI(uri)
+	if err != nil {
+		return fmt.Errorf("the URI provided: %s is invalid", uri)
+	}
+	if parsedURL.Scheme == "http" || parsedURL.Scheme == "https" {
+		var sha256Checksum string
+		if sha256Checksums, ok := parsedURL.Query()["sha256"]; ok {
+			sha256Checksum = sha256Checksums[0]
+		}
+		if sha256Checksum == "" {
+			return fmt.Errorf("the sha256 parameter in the %s URI is missing", uri)
 		}
+		if len(sha256Checksum) != 64 {
+			return fmt.Errorf("the sha256 parameter in the %s URI is invalid", uri)
+		}
+	} else {
+		return fmt.Errorf("the URI provided: %s must begin with http/https", uri)
+	}
+	if !strings.HasSuffix(parsedURL.Path, ".qcow2.gz") && !strings.HasSuffix(parsedURL.Path, ".qcow2.xz") {
+		return fmt.Errorf("the URI provided: %s is an OS image and must end in .qcow2.gz or .qcow2.xz", uri)
 	}
 	return nil
 }
 
-func validateDisabledConfig(prov *metal3iov1alpha1.Provisioning) error {
-	for _, toTest := range []struct {
-		Name  string
-		Value string
-	}{
-
-		{Name: "ProvisioningIP", Value: prov.Spec.ProvisioningIP},
-		{Name: "ProvisioningNetworkCIDR", Value: prov.Spec.ProvisioningNetworkCIDR},
-		{Name: "ProvisioningOSDownloadURL", Value: prov.Spec.ProvisioningOSDownloadURL},
-	} {
-		if toTest.Value == "" {
-			return fmt.Errorf("%s is required but is empty", toTest.Name)
+func validateProvisioningConfig(pc map[string]string) error {
```

Comment:
Is there a way for us to build a list of all of the errors and return them all?

Reply:
yeah that's a good idea.  I'd have to check if webhooks is cool with a list but otherwise it'd be fine. I guess we could also just return a longer error string if not.

Reply:
Yeah, if we return a list the caller can do what it wants with the list, including formatting a response payload as a single string if that's all it supports.

Reply:
sounds good